### DISCUSSION TITLE
Jaccard Similarity Index Function

### DISCUSSION THREAD

**Discussion by gramian** [ArcadeData/arcadedb#625] [Ideas]
Dear Developers,
I would like to propose a new function to determine similarity of sets or collections: the [Jaccard similarity index](https://en.wikipedia.org/wiki/Jaccard_index). Such a function is also available in other databases like [AsterixDB](https://asterixdb.apache.org/docs/0.9.4.1/sqlpp/builtins.html#similarity_jaccard) or [ArangoDB](https://www.arangodb.com/docs/stable/aql/functions-array.html#jaccard). Given the functions `intersect()`, `unionall()`, `count()` and the method `.asFloat()`such a function can be emulated in ArcadeDB's SQL by something like:
```
SELECT count(intersect($a, $b)).asFloat() / count(unionall($a, $b)).asFloat() AS jaccard;
```
given sub-queries `$a` and `$b`. However, this is only convenient as an end result, but not as an intermediate step, for example, if based on the similarity something else should happen. Thus, IMHO this function would be a good addition of functionality ArcadeDB currently does not provide.

Thank you for considering this proposition

**Comment by gramian** [comment]
Here is a Jaccard similarity index in ArcadeDB's SQL I am [currently using](https://github.com/gramian/chicken-arcadedb):
```
LET $t = SELECT unionall(x) AS x,
                unionall(y) AS y FROM mytype;
SELECT intersect($t[0].x,$t[0].y).size().asFloat() /
        unionall($t[0].x,$t[0].y).size().asFloat();
```
given two fields `x`, `y` and a type `mytype`.

**Reply by gramian** [reply]
Here is an even simpler (one liner, aka `sql` instead of `sqlscript`) Jaccard similarity index:
```
SELECT intersect($t[0].x,$t[0].y).size().asFloat() / unionall($t[0].x,$t[0].y).size().asFloat()
LET $t = (SELECT unionall(x) AS x, unionall(y) AS y FROM mytype);
```

**Comment by lvca** [comment]
I'd love to add a new "Vector" model in ArcadeDB and Jaccard together to cosine and others could be written as native functions to increase performance and easy to use. WDYT?

**Reply by gramian** [reply]
I like a `jaccard` function, but note it is aimed at sets not vectors.

**Reply by lvca** [reply]
I see it's also used with strings: https://commons.apache.org/sandbox/commons-text/jacoco/org.apache.commons.text.similarity/JaccardSimilarity.java.html/
